<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Исследование-данных" data-toc-modified-id="Исследование-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Исследование данных</a></span></li><li><span><a href="#Анализ-объектов,-содержащихся-в-базе-данных-сервиса" data-toc-modified-id="Анализ-объектов,-содержащихся-в-базе-данных-сервиса-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Анализ объектов, содержащихся в базе данных сервиса</a></span><ul class="toc-item"><li><span><a href="#Количество-книг,-которые-вышли-после-1-января-2000-года" data-toc-modified-id="Количество-книг,-которые-вышли-после-1-января-2000-года-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Количество книг, которые вышли после 1 января 2000 года</a></span></li><li><span><a href="#Количество-обзоров-и-средняя-оценка-для-каждой-книги" data-toc-modified-id="Количество-обзоров-и-средняя-оценка-для-каждой-книги-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Количество обзоров и средняя оценка для каждой книги</a></span></li><li><span><a href="#Издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц" data-toc-modified-id="Издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Издательство, которое выпустило наибольшее число книг толще 50 страниц</a></span></li><li><span><a href="#Автор-с-самой-высокой-средней-оценкой-книг-(учтены-только-книги-с-50-и-более-оценками)" data-toc-modified-id="Автор-с-самой-высокой-средней-оценкой-книг-(учтены-только-книги-с-50-и-более-оценками)-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Автор с самой высокой средней оценкой книг (учтены только книги с 50 и более оценками)</a></span></li><li><span><a href="#Cреднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок" data-toc-modified-id="Cреднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Cреднее количество обзоров от пользователей, которые поставили больше 50 оценок</a></span></li></ul></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Выводы</a></span></li></ul></div>

# Анализ базы данных сервиса для чтения книг

В предостоящем проекте проведем исследование данных, содержащихся в сервисе для чтения книг.

Цель исследования — проанализировать базу данных сервиса, содержащую информацию о его различных объектах, которая поможет сформулировать ценностное предложение для нового продукта.

Анализ данных пройдёт по следующим направлениям:

- Обзор исходных данных
- Анализ объектов, содержащихся в базе данных сервиса (книги и их авторы, пользовательские обзоры и рейтинги, а также издательства)
- Вывод

# Введение 

In [74]:
# импортируем библиотеки

import pandas as pd
from sqlalchemy import create_engine

In [75]:
# устанавливаем параметры

db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])

In [77]:
# сохраняем коннектор

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

## Исследование данных

In [2]:
def data(i):
    return pd.io.sql.read_sql(i, con = engine)

In [3]:
query_books = '''

SELECT * 
FROM books
LIMIT 5;

'''
books = pd.io.sql.read_sql(query_books, con = engine)

query_authors = '''
SELECT * 
FROM authors
LIMIT 5;
'''

authors = pd.io.sql.read_sql(query_authors, con = engine)

query_publishers = '''

SELECT * 
FROM publishers
LIMIT 5;

'''

publishers = pd.io.sql.read_sql(query_publishers, con = engine)

query_ratings = '''

SELECT * 
FROM ratings
LIMIT 5;

'''

ratings = pd.io.sql.read_sql(query_ratings, con = engine)

query_reviews = '''

SELECT * 
FROM reviews
LIMIT 5;

'''
reviews = pd.io.sql.read_sql(query_reviews, con = engine)

In [4]:
data(query_books)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

В датасете books содержится информация с названием книги, дате ее публикации, а также id книги, id издательства и id автора.

In [5]:
data(query_authors)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

Таблица authors включает в себя данные об авторах: их имена и id.

In [6]:
data(query_publishers)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

publishers содержит данные об издательствах: названия издательств и их id.

In [7]:
data(query_ratings)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

Датасет ratings позволяет узнать о пользователях и об оценках, которые они ставят книгам: имя пользователя, id оценки, id книги и оценка книги.

In [8]:
data(query_reviews)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

query_reviews содержит информацию о пользовательских обзорах и включает в себя следующие позици: id обзора, id книги, имя пользователя и текст обзора.

Таким образом, выведены первые строки исследуемых датасетов, содержащие следующую информацию о книгах, их авторах, издательствах, пользовательских оценках книг и пользовательских обзоров.

## Анализ объектов, содержащихся в базе данных сервиса 

### Количество книг, которые вышли после 1 января 2000 года

In [9]:
query_amount_book = '''

SELECT COUNT(book_id) AS amount
FROM books
WHERE publication_date > '2000-01-01'

'''
amount_book = pd.io.sql.read_sql(query_amount_book, con = engine)
amount_book

amount
0     819

### Количество обзоров и средняя оценка для каждой книги 

In [78]:
query_ratings_reviews = '''

SELECT b.book_id,
       b.title,
       COUNT(DISTINCT(r.review_id)) AS reviews_amount,
       ROUND(AVG(ra.rating),2) AS avg_rating
FROM books AS b
LEFT JOIN reviews AS r ON r.book_id = b.book_id
LEFT JOIN ratings AS ra ON ra.book_id = b.book_id
GROUP BY b.book_id, 
         b.title
ORDER BY reviews_amount DESC;

'''
ratings_reviews = pd.io.sql.read_sql(query_ratings_reviews, con = engine)
ratings_reviews

book_id                                              title  \
0        948                            Twilight (Twilight  #1)   
1        963                                Water for Elephants   
2        734                                   The Glass Castle   
3        302  Harry Potter and the Prisoner of Azkaban (Harr...   
4        695  The Curious Incident of the Dog in the Night-Time   
..       ...                                                ...   
995       83    Anne Rice's The Vampire Lestat: A Graphic Novel   
996      808                            The Natural Way to Draw   
997      672   The Cat in the Hat and Other Dr. Seuss Favorites   
998      221                          Essential Tales and Poems   
999      191  Disney's Beauty and the Beast (A Little Golden...   

     reviews_amount  avg_rating  
0                 7        3.66  
1                 6        3.98  
2                 6        4.21  
3                 6        4.41  
4                 6        4.08  
..              ...         ...  
995               0        3.67  
996               0        3.00  
997               0        5.00  
998               0        4.00  
999               0        4.00  

[1000 rows x 4 columns]

### Издательство, которое выпустило наибольшее число книг толще 50 страниц

In [11]:
query_publisher = '''

SELECT p.publisher_id,
       p.publisher,
       COUNT(b.book_id) AS amount_books
       
FROM publishers AS p
INNER JOIN books AS b ON p.publisher_id = b.publisher_id
WHERE b.num_pages > 50
GROUP BY p.publisher_id,
       p.publisher
ORDER BY amount_books DESC
LIMIT 1;

'''
publisher_books = pd.io.sql.read_sql(query_publisher, con = engine)
publisher_books

publisher_id      publisher  amount_books
0           212  Penguin Books            42

### Автор с самой высокой средней оценкой книг (учтены только книги с 50 и более оценками)

In [16]:
query_popular_author = '''

SELECT author,
       ROUND(AVG(popular_books.avg_rating),2) AS avg_rating
FROM (SELECT a.author,
             b.book_id,
             AVG(r.rating) AS avg_rating,
             COUNT(r.rating) AS rating_amount
      FROM authors AS a
      INNER JOIN books AS b ON b.author_id = a.author_id
      INNER JOIN ratings AS r ON r.book_id = b.book_id
      GROUP BY a.author,
            b.book_id
      HAVING COUNT(r.rating) >= 50) AS popular_books
GROUP BY author
ORDER BY avg_rating DESC    
LIMIT 1;

'''
popular_author = pd.io.sql.read_sql(query_popular_author, con = engine)
popular_author

author  avg_rating
0  J.K. Rowling/Mary GrandPré        4.28

### Cреднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [13]:
query_avg_reviews_amount = '''

SELECT ROUND(AVG(count)) AS avg_amount
FROM (SELECT username,
             COUNT(review_id)
      FROM reviews
      WHERE username IN (SELECT username
                         FROM ratings
                         GROUP BY username
                         HAVING COUNT(rating_id) > 50)
      GROUP BY username)  AS count_reviews

'''
avg_reviews_amount = pd.io.sql.read_sql(query_avg_reviews_amount, con = engine)
avg_reviews_amount

avg_amount
0        24.0

## Выводы

Проанализировав данные об объектах базы данных сервиса, сформулируем следующие выводы:
    
- В базе данных содержится 819 книг, которые вышли после 1 января 2000 года
- Таблица ratings_reviews включает в себя количество обзоров и средняя оценка для каждой книги: наибольшее количество обзоров определено у книги Twilight (Twilight #1)
- "Penguin Books" - издательство, которое выпустило наибольшее число книг толще 50 страниц: количество книг составляет 42 штуки
- J.K. Rowling, работающая над книгами вместе с иллюстратором Mary GrandPré, является автором с самой высокой средней оценкой книг - 4.28 (при расчете учитывались только книги с 50 и более оценками)
- 24 - среднее количество обзоров от пользователей, которые поставили больше 50 оценок
    